\chapter{Week 4: Diepere duik}
Deze week gaan kijken naar enkele concepten die ons leven als programmeur een stukje makkelijker kunnen gaan maken. We gaan eerst een blik werpen op verzamelingen in de vorm van \textit{lijsten}. Daarna de robuustheid van onze code aanzienlijk verbeteren met \pyth{Try} / \pyth{Except}-blokken en ten slotte meer structuur aanbrengen met de introductie van \textit{functies}. 
% , \textit{Dictionaries} en \textit{tuples}

\section{Lijsten}\index{Lijsten}
De lijsten in \textit{Python} zijn vergelijkbaar met de array's in \textit{C}, maar dan veel flexibeler in te zetten. Het voordeel van een lijst te gebruiken, is dat je data die bij elkaar hoort mooi kunt groeperen. Maar je kunt ook handige operaties op een lijst uitvoeren, zo kun je 'm bijvoorbeeld sorteren, omkeren, of elementen eruit filteren. Eerst gaan we kijken naar het aanmaken ervan, hieronder volgen een aantal willekeurige lijsten:

\begin{python}
# Lijsten met getallen:
cijfers = [1, 2, 3, 5, 10, 15] 
frac = [0.5, 0.25, 0.2, 0.1, 0.001, 0.0005] 

# Lijsten met strings:
namen = ['Piet', 'Jan', 'Klaas']
dagen = ['ma', 'di', 'wo', 'do', 'vr', 'za', 'zo']

# Een lijst met verschillende datatypes in een, kan ook:
mix = [13, 'robotica', 3, 2, 'analyse', 3.14, cijfers]
\end{python}

Lijsten hebben zoals je hierboven kunt zien een naam en een verzameling aan elementen. De elementen staan tussen de blokhaken $[\ ]$, en zijn gescheiden door een komma. Lijsten kunnen elk data type bevatten (er kunnen zelfs verschillende datatypes - en dus ook weer lijsten - in een lijst staan, wat we verder in dit vak niet meer gaan behandelen).

\newpage

Lijsten zijn bijv. heel krachtige dingen om in te zetten in combinatie met \pyth{for}-loops:
\begin{python}
kleuren = ['geel', 'rood', 'groen', 'oranje']

for kleur in kleuren:
	print(f'He, nu issie weer {kleur}!')
\end{python}

In deze \pyth{for}-loop worden alle elementen van \textit{kleuren} een voor een opgeslagen in de variabele \textit{kleur} (regel $3$). En die wordt dan weer gebruikt op regel $4$. De output van het bovenstaande stukje code is als volgt:

\begin{python}
He, nu issie weer geel
He, nu issie weer rood
He, nu issie weer groen
He, nu issie weer oranje
\end{python}

Zoals is gezegd, zijn er een aantal handige operaties uit te voeren op een lijst:
\begin{python}
>>> a = [3,2,4,1]
>>> len(a)
4
>>> a.append(5)
>>> a
[3, 2, 4, 1, 5]
>>> len(a)
5
>>> a.reverse()
>>> a
[5, 1, 4, 2, 3]
>>> a.sort()
>>> a
[1, 2, 3, 4, 5]
\end{python}

\begin{exercise}
Maak zelf een lijst aan. Ga na wat de functies \pyth{count()}, \pyth{pop()} en \pyth{remove()} doen. 
\end{exercise}

% \section{Tuples}\index{Tuples}

% \section{Dictionaries}\index{Dictionaries}

\section{Try ... Except}\index{Try ... Except}
Je bent ze inmiddels ongetwijfeld tijdens het maken van de huiswerkopdrachten (of elders) tegengekomen: \textit{errors}. Bijv. bij het opvragen van gegevens bij de gebruiker, via de \pyth{input()}-functie:
\begin{python}
x = input('Geef een cijfer: ')
x = int(x)  # Zet de input-string om naar een int, en sla dit weer op in 'x'.
x += 1  # Tel er 1 bij op
print(f'Een hoger: {x}')
\end{python}
Deze code runt doorgaans prima:
\begin{python}
Geef een cijfer: 12
Een hoger: 13
\end{python}
Maar wat nu als de gebruiker iets verkeerds intypt? Bijv. een letter:
\begin{python}
Geef een cijfer: q
Traceback (most recent call last):
  File "/Users/stefan/Code/Python/err_test.py", line 2, in <module>
    x = int(x)  # Zet de input-string om naar een int, en sla dit weer op in 'x'.
ValueError: invalid literal for int() with base 10: 'q'
\end{python}
Dan krijg je dus een \textit{error} voor je kiezen, in dit geval een \pyth{ValueError}. Bijkomend nadeel: het programma sluit direct af (crasht). Gelukkig geeft \pyth{Python} bij errors vaak genoeg informatie waarmee je de bug je in je code kunt oplossen. Hierboven kun je lezen dat er iets mis gaat op regel $2$: \pyth{x = int(x)}, op het moment dat we de \pyth{str} $q$ om proberen te zetten naar een \pyth{int}. En dat is best te verklaren natuurlijk, $q$ is domweg geen getal. 

\begin{remark}
\textbf{Tip: } Mocht je nu op zoek zijn naar de \textit{ASCII}-waarde van een karakter, gebruik dan de \pyth{ord()}-functie:
\begin{python}
>>> ord('a')
97
>>> ord('b')
98
\end{python}
\end{remark}

Een gebruiker kan nou eenmaal iets verkeerds intypen en het zou vrij suf zijn als dan elke keer je programma crasht. Vandaar dat er in \textit{Python} functionaliteit zit om dit soort fouten van buitenaf af te vangen en op te reageren. Dit doen we door onze kritische code (in dit geval het omzetten van een \pyth{str} naar een \pyth{int}: \pyth{x = int(x)}) in een \pyth{try} / \pyth{except} blok te zetten. En dat ziet er als volgt uit:

\begin{python}
x = input('Geef een cijfer: ')
try:
	x = int(x)  # Zet de input-string om naar een int, en sla dit weer op in 'x'.
	x += 1  # Tel er 1 bij op
	print(f'Een hoger: {x}')
except ValueError:
    print('Dat was geen getal, probeer het nog eens')
\end{python}

Hier 'proberen' we de \pyth{str} te interpreteren als een \pyth{int}. Gaat dat goed, dan wordt er netjes de rest van de code uitgevoerd. Krijg je een error, dan wordt het stuk code na de \pyth{except} uitgevoerd, en wordt de gebruiker vriendelijk verzocht 'even normaal te doen' en het nog eens te proberen. \newline

Elke error die je tegenkomt kun je op deze manier afvangen. Let daar wel bij op, dat niet elke error afgevangen hoeft te worden. Fouten die je zelf als programmeur maakt horen daar vaak niet bij, het gaat meer om fouten van 'buitenaf'. Bijv. je probeert connectie te maken met een server, maar die is op het moment uit de lucht. Dan zul je wellicht iets in de trant van een \pyth{ConnectionError} of een \pyth{TimeOutError} krijgen, die zijn handig om af te vangen. Je wilt niet dat door iets van buitenaf je programma crasht. Voor de liefhebber: hier is een lijstje te zien van de errors die standaard te vinden zijn in \textit{Python}: \url{https://docs.python.org/3/library/exceptions.html}.

\newpage

\section{Functies}\index{Functies}
\textit{Functies} waren we ook al tegengekomen in \textit{C}, in \textit{Python} is de opzet ervan hetzelfde: Een blok aan code, die alleen wordt uitgevoerd als de functie wordt 'aangeroepen'. Je maakt ze aan ('definieert' ze) met \pyth{def}, zie ook het onderstaande voorbeeld:

\begin{python}
def my_function():
    print('Hallo vanuit een functie!') 

my_function()
print('Hallo vanuit het hoofdprogramma!')
my_function()
\end{python}

Dit produceert de volgende output:

\begin{python}
Hallo vanuit een functie!
Hallo vanuit het hoofdprogramma!
Hallo vanuit een functie!
\end{python}

Je ziet in het programma op de eerste $2$ regels de gemaakte functie, na het woordje \pyth{def}. De naam van de functie is \textit{my\_function}, alle regels die daarna volgen en ingesprongen zijn (tab), horen bij deze functie. 
In dit geval is dit maar $1$ regel, namelijk een \pyth{print()}. \newline

Daarna wordt deze functie 'aangeroepen', dat gebeurt voor het eerst op regel $4$. Dat doe je dus door de naam van de functie te typen, gevolgd door $2$ haakjes $( )$. Even later op regel $6$ gebeurt hetzelfde nog eens. \newline

Dit was een voorbeeld van een functie zonder argumenten. Vaak zul je ook functies zien en maken die dat juist wel hebben. Een argument is iets wat je meegeeft met de functie. Dat kunnen er zoveel zijn als je wilt, en ook van elk datatype. Hieronder een voorbeeld met $2$ argumenten, van het type \pyth{str}:

\begin{python}
def my_function(naam, vraag):
    print(f'Hallo {naam}! {vraag}?') 

my_function('Henk', 'Hoe gaat het')
my_function('Piet', 'Waddup')
\end{python}
De uitoer zal in dit geval zijn:
\begin{python}
Hallo Henk! Hoe gaat het?
Hallo Piet! Waddup?
\end{python}

Valt je op je dat nog steeds nergens het datatype hoeft aan te geven? \textit{Python} snapt dat het om \pyth{str} gaat, omdat je ze als zodanig meegeeft op regel $4$ en $5$. 

\begin{remark}
De functie zal het ook prima doen als je 'm aanroept met bijv. getallen: \newline 
\pyth{my_fucntion(1, 2)}
\begin{python}
Hallo 1! 2?
\end{python}
\textit{Python} gaat heel flexibel met om datatypes. Omdat alles wat in de functie \textit{my\_function} gebeurt met de argumenten \textit{naam} en \textit{vraag} ook prima gedaan kan worden met getallen, krijg je geen errors. In \textit{C} vlogen nu de errors en warnings om je oren.
\end{remark}

\newpage

Dan rest ons nog een ding om in te duiken, functies die iets teruggeven. Het kan zijn dat een functie een bepaalde operatie uitvoerd, en dat je het resultaat daarvan graag wil gebruiken verderop in je programma. Dat kan, net als in \textit{C} met \pyth{return}. Zie ook het volgende voorbeeld:

\begin{python}
import math 

def oppervlakte(r):
    # Berekend de oppervlakte van een cirkel ahv de straal.
    opp = r**2 * math.pi
    return opp

o5 = oppervlakte(r=5)
print(f'Oppervlakte cirkel, bij straal=5:  {o5}')

o10 = oppervlakte(10)
print(f'Oppervlakte cirkel, bij straal=10: {o10}')

print(f'Oppervlakte cirkel, bij straal=15: {oppervlakte(15)}')

o20 = oppervlakte(20)
print(f'Oppervlakte cirkel, bij straal=20: {o20:.2f}')
\end{python}

In tegenstelling tot \textit{C} hoef je ook hier geen datatype op te geven, dat wordt voor je geregeld. Enkel de functie afsluiten met een \pyth{return}-statement is voldoende om de functie een waarde terug te laten geven. \newline


De functie \pyth{oppervlakte()} berekend de oppervlakte van een cirkel a.h.v. de meegegeven straal \pyth{r}. En hij wordt op vier verschillende manieren aangeroepen:
\begin{enumerate}
	\item[-] Allereerst op regel $8$ met een straal van $r=5$, hier wordt ook de naam van het argument ($r$) meegegeven, dat leest makkelijker, maar is niet nodig. Het antwoord van de functie wordt opgeslagen in variabele $o5$ en naar het scherm gescreven. \newline

	\item[-] Daarna wordt de functie aangeroepen met een straal van $r=10$, het antwoord hiervan wordt daarna opgeslagen in de variabele $o10$, die dan geprint wordt op het scherm. \newline

	\item[-] Daarna op regel $14$, wordt het aanroepen van de functie (met $r=15$) en het printen van de teruggegeven waarde gecombineerd tot een regel code. \newline

	\item[-] En tenslotte wordt de functie nog een keer aangeroepen voor $r=20$ op regel $16$. Bij het printen (regel $17$) gebeurt iets bijzonders, daar wordt namelijk aangegeven dat de uitvoer maar $2$ kommagetallen moet laten zien (dit komt door de \pyth{:.2f} in \pyth{print}). 
\end{enumerate}

De uitvoer van het programma is dan ook:
\begin{python}
Oppervlakte cirkel, bij straal=5:  78.53981633974483
Oppervlakte cirkel, bij straal=10: 314.1592653589793
Oppervlakte cirkel, bij straal=15: 706.8583470577034
Oppervlakte cirkel, bij straal=20: 1256.64
\end{python}

\newpage
Zoals ik vorige week stelde, kun je met functies in combinatie met de \pyth{gpiozero}-module leuke dingen doen. Zo kun je dus bijv. functies aanroepen, als er een bepaalde actie wordt uitgevoerd word, zoals het indrukken en het loslaten van een knop:
\begin{python}
from gpiozero import Buttons
from signal import pause

def ingedrukt():
	print('knop ingedrukt!')

def losgelaten():
	print('knop weer losgelaten!')

btn = Button(2)  # Drukknop zit op GPIO2

btn.when_pressed = ingedrukt    # Koppel functie bij het indrukken.
btn.when_released = losgelaten  # Koppel functie bij het loslaten.
pause()                         # Doe verder niets meer, maar sluit niet af.
\end{python}

\begin{exercise}
Voor het programma uit, en ga voor jezelf na wat er precies gebeurt.
\end{exercise}

\begin{remark}
Weet je nog uit Programmeren $1$, dat knoppen last kunnen hebben van het zogeheten \textit{bouncen}, waardoor het lijkt dat ze veel vaker ingedrukt worden, dan ze daadwerkelijk worden. Mocht je daar nu ook last van hebben, dan kun je met \pyth{gpiozero} vrij makkelijk de bounce-tijd instellen. Vervang regel $10$ door:
\begin{python}
# Drukknop zit op GPIO2 met 100ms bounce time:
btn = Button(2, bounce_time=0.1)  
\end{python}
\end{remark}

De laatste regel \pyth{pause()}, zorgt dat het programma verder niets meer uitvoert, (behalve het afhandelen van de knop) maar ook niet afsluit. In tegenstelling tot de \pyth{while True:} uit het vorige hoofdstuk, waardoor de \textit{Raspberry Pi} intensief gebruikt werd, kan hij in dit geval rustig andere taken doen. 

\begin{remark}
Het mooie van het kunnen koppelen van functies aan gebeurtenissen van een sensor (zoals hier in het geval van de drukknop), is dat dat ook functies kunnen zijn die gebruikt worden door actuatoren, zoals bijv. een LED. Een LED heeft een \pyth{on()}- en een \pyth{off()}-functie. Dus het koppelen van een LED aan een drukknop kan zo:

\begin{python}
from gpiozero import Button, LED
from signal import pause

btn = Button(2)  # Drukknop zit op GPIO2
led = LED(17)    # LED zit op GPIO17

# Koppel het indrukken van de knop met led.on():
btn.when_pressed = led.on    
# Koppel het loslaten van de knop met led.off():
btn.when_released = led.off  

# Doe verder niets meer, maar sluit niet af:
pause()                      
\end{python}
\end{remark}

\newpage
Afsluitend aan dit hoofdstuk gaan we eens wat dingen combineren die we geleerd hebben. Je kunt namelijk elementen uit een lijst filteren, als je daarvoor een filter-functie voor aanmaakt. Bijv. je hebt een bepaalde temperatuursensor, waar je elk uur de temperatuur mee meet, maar deze geeft af en toe een foute waarde (bijv. +$100^\circ C$ op een winterdag). 

\begin{python}
def temp_filter(temp):
    # Filter-functie, voor temp groter dan 90 graden.
    if (temp > 90):
        return False
    else:
        return True

# Lijst met gemeten temperatuur in celcius:  
gemeten_temp = [15.2, 14.4, 14.2, 99.9, 13.8, 12.6, 100.1]

# Pas filter toe:
gefilterde_temp = filter(temp_filter, gemeten_temp)

# Print de overgebleven waardes op het scherm:
print('De gefilterde temperaturen zijn:')
for t in gefilterde_temp:
    print(t)                    
\end{python}

In het bovenstaande voorbeeld, wordt een filter-functie aangemaakt \pyth{temp_filter(temp)}. Die een \pyth{False} teruggeeft als de meegegeven temperatuur \pyth{temp} groter is dan $90$, en anders \pyth{True}. \newline

Op regel $12$ wordt die filterfunctie daadwerkelijk toegepast op de lijst \pyth{gemeten_temp} en de nieuwe lijst die dat opleverd wordt opgelsagen in \pyth{gefilterde_temp}

\newpage

\section{Huiswerkopdrachten}\index{Huiswerkopdrachten}

\begin{exercise}\label{exc6:exc1}
Maak een lege lijst aan, en vul deze daarna met de getallen $1$ t/m $100$.
Gebruik hiervoor een loop en de \pyth{append()} functie. \newline
\textbf{Tip:} Een lege lijst maak je als volgt aan: \pyth{lijst = []}
\end{exercise}

\begin{exercise}
Maak een filter-functie waarmee je alle getallen onder de $30$ uit een lijst kunt filteren. Pas deze toe op de lijst die je gemaakt hebt bij opdracht \ref{exc6:exc1}.
\end{exercise}

\begin{exercise}
Maak een filter-functie waarmee je oneven getallen uit een lijst kunt filteren. Pas deze toe op de lijst die je gemaakt hebt bij opdracht \ref{exc6:exc1}.
\end{exercise}

\begin{exercise}
\label{exc6:exc4}
Maak een functie \pyth{schreeuw} die een meegegeven str omzet naar hoofdletters, en deze daarna weer terug geeft met \pyth{return}. 
\end{exercise}

\begin{exercise}
Wat als je nu de functie die net gemaakt hebt bij \ref{exc6:exc4}, aanroept met een getal (bijv. \pyth{schreeuw(5)})? \newline
Breidt de functie uit met een \pyth{try} / \pyth{except}, zodat deze niet langer crasht. 
\end{exercise}

\begin{exercise}
Sluit $4$ LEDs aan op de \textit{GPIO17}, \textit{GPIO18}, \textit{GPIO3} en \textit{GPIO5} pins. Configureer ze, en zet ze daarna in een lijst. Ga met een loop door de lijst heen, en zet ze een voor een $1$ seconde aan, en dan weer uit.
\end{exercise}

\begin{exercise}
Maak een programma dat $10$ keer aan de gebruiker vraagt om een dier in te voeren. Voeg elk van deze dieren toe aan een lijst. Filter alle diernamen die minder dan $3$ letters hebben en print daarna de lijst in alfabetische volgorde op het scherm. \newline
\textbf{Tip:} Je bepaalt de lengte van een string (of lijst) met de functie: \pyth{len(mijn_string)}.
\end{exercise}
